src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -383,249 +507,339 @@ private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
+            => new Mapper(this, Schema.Create(schema));
+
+        private sealed class Mapper : MapperBase
         {
-            long? rows = Source.GetRowCount(lazy: false);
-            if (rows != null)
-                return rows.GetValueOrDefault();
+            private readonly WhiteningTransform _parent;
+            private readonly int[] _cols;
+            private readonly ColumnType[] _srcTypes;
 
-            int maxRows = _exes.Max(i => i.MaxRow);
-            long r = 0;
-            using (var cursor = Source.GetRowCursor(col => false))
+            public Mapper(WhiteningTransform parent, Schema inputSchema)
+                : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                while (r < maxRows && cursor.MoveNext())
-                    r++;
-            }
-            return r;
-        }
+                _parent = parent;
+                _cols = new int[_parent.ColumnPairs.Length];
+                _srcTypes = new ColumnType[_parent.ColumnPairs.Length];
 
-        protected override ColumnType GetColumnTypeCore(int iinfo)
-        {
-            Host.Check(0 <= iinfo & iinfo < Infos.Length);
-            return _exes[iinfo].Type;
-        }
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    var info = _parent._infos[i];
+                    if (!InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out _cols[i]))
+                        throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", _parent.ColumnPairs[i].input);
+                    _srcTypes[i] = inputSchema.GetColumnType(_cols[i]);
+                    ColumnType outType = (info.Kind == WhiteningKind.Pca && info.PcaNum > 0) ? new VectorType(NumberType.Float, info.PcaNum) : _srcTypes[i];
+                    ValidateModel(Host, _parent._models[i], outType);
+                    if (info.SaveInv)
+                        ValidateModel(Host, _parent._invModels[i], outType);
+                }
+            }
 
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
-        {
-            Host.AssertValueOrNull(ch);
-            Host.AssertValue(input);
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            disposer = null;
-
-            var ex = _exes[iinfo];
-            Host.Assert(ex.Kind == WhiteningKind.Pca || ex.Kind == WhiteningKind.Zca);
-            var getSrc = GetSrcGetter<VBuffer<Float>>(input, iinfo);
-            var src = default(VBuffer<Float>);
-            int cslotSrc = Infos[iinfo].TypeSrc.ValueCount;
-            int cslotDst = (ex.Kind == WhiteningKind.Pca && ex.PcaNum > 0) ? ex.PcaNum : Infos[iinfo].TypeSrc.ValueCount;
-            var model = _models[iinfo];
-            ValueGetter<VBuffer<Float>> del =
-                (ref VBuffer<Float> dst) =>
+            public override Schema.Column[] GetOutputColumns()
```

Comment:
>public [](start = 12, length = 6)

Is this needed to be public?

Reply:
Yes that's how it is defined in the mapper base class.